File: dotnet/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example4/ChatUser.cs
namespace MediatorDesignPattern.Example4;

/// <summary>
/// Concrete Colleague - Chat participant
/// </summary>
public class ChatUser : User
{
    public ChatUser(IChatMediator mediator, string name) 
        : base(mediator, name)
    {
    }

    public override void Send(string message)
    {
        Console.WriteLine($"[{Name}] Sending: {message}");
        Mediator.SendMessage(message, this);
    }

    public override void SendPrivate(string message, User receiver)
    {
        Console.WriteLine($"[{Name}] Sending private to {receiver.Name}: {message}");
        Mediator.SendPrivateMessage(message, this, receiver);
    }

    public override void Receive(string message, User sender)
    {
        Console.WriteLine($"[{Name}] Received from {sender.Name}: {message}");
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example4/src/ChatUser.ts
import { User } from './User';
import { IChatMediator } from './IChatMediator';

export class ChatUser extends User {
  constructor(mediator: IChatMediator, name: string) {
    super(mediator, name);
  }

  send(message: string): void {
    console.log(`[${this.name}] Sending: ${message}`);
    this.mediator.sendMessage(message, this);
  }

  sendPrivate(message: string, receiver: User): void {
    console.log(`[${this.name}] Sending private to ${receiver.name}: ${message}`);
    this.mediator.sendPrivateMessage(message, this, receiver);
  }

  receive(message: string, sender: User): void {
    console.log(`[${this.name}] Received from ${sender.name}: ${message}`);
  }
}
